media_driver/linux/common/os/i915/mos_bufmgr.c

```diff
@@ -1678,7 +1838,51 @@ drm_export int mos_gem_bo_map(struct mos_linux_bo *bo, int write_enable)
 
     pthread_mutex_lock(&bufmgr_gem->lock);
 
-    if (!bo_gem->mem_virtual) {
+    if (!bo_gem->mem_virtual && bufmgr_gem->has_lmem) {
+        struct drm_i915_gem_mmap_offset mmap_arg;
+
+        MOS_DBG("bo_map: %d (%s), map_count=%d\n",
+            bo_gem->gem_handle, bo_gem->name, bo_gem->map_count);
+
+        memclear(mmap_arg);
+        mmap_arg.handle = bo_gem->gem_handle;
+        mmap_arg.flags = I915_MMAP_OFFSET_FIXED;
+        ret = drmIoctl(bufmgr_gem->fd,
+                   DRM_IOCTL_I915_GEM_MMAP_OFFSET,
+                   &mmap_arg);
+        if (ret != 0) {
+            ret = -errno;
+            MOS_DBG("%s:%d: Error mapping buffer %d (%s): %s .\n",
+                __FILE__, __LINE__, bo_gem->gem_handle,
+                bo_gem->name, strerror(errno));
+            pthread_mutex_unlock(&bufmgr_gem->lock);
+            return ret;
+        }
+
+        /* and mmap it */
+        bo_gem->mem_virtual = drm_mmap(0, bo->size, PROT_READ | PROT_WRITE,
+                           MAP_SHARED, bufmgr_gem->fd,
+                           mmap_arg.offset);
+        if (bo_gem->mem_virtual == MAP_FAILED) {
+            bo_gem->mem_virtual = nullptr;
+            ret = -errno;
+            MOS_DBG("%s:%d: Error mapping buffer %d (%s): %s .\n",
+                __FILE__, __LINE__,
+                bo_gem->gem_handle, bo_gem->name,
+                strerror(errno));
+        }
+
+        assert(bufmgr_gem->has_wait_timeout);
+        memclear(wait);
+        wait.bo_handle = bo_gem->gem_handle;
+        wait.timeout_ns = -1; // infinite wait
+        ret = drmIoctl(bufmgr_gem->fd, DRM_IOCTL_I915_GEM_WAIT, &wait);
+        if (ret == -1) {
+            MOS_DBG("%s:%d: DRM_IOCTL_I915_GEM_WAIT failed (%d)\n",
+                __FILE__, __LINE__, errno);
+        }
```

Comment:
Still need to wait when bo is already mapped.
if (has_lmem)
{
  if (!bo_gem->mem_virtual)
  {
    map;
  }
  wait;
}
else /*!has_lmem*/
{
  if (!bo_gem->mem_virtual)
  {
    map;
  }
  set_domain;
}